From b7a43f021393e72b3f3ae54405453f6512a72850 Mon Sep 17 00:00:00 2001
From: Aaron Kling <webgeek1234@gmail.com>
Date: Thu, 31 Dec 2015 22:33:50 -0600
Subject: [PATCH 4/4] Add vendor hook to handle_control_message

This is needed for Nvidia Shield devices to handle a
'restart consolemode' request from a blobbed stock app.

The vendor function is expected to return non-0 if it did
not handle the message and 0 if it did (or wants it ignored).

Change-Id: Ia8e4ba0bbf561f29f72862cd986f1660d7b501da
---
 init/init.cpp        | 4 ++++
 init/vendor_init.cpp | 7 +++++++
 init/vendor_init.h   | 3 +++
 3 files changed, 14 insertions(+)

diff --git a/init/init.cpp b/init/init.cpp
index 35fc442d0..e13eac5b2 100644
--- a/init/init.cpp
+++ b/init/init.cpp
@@ -68,6 +68,7 @@
 #include "ueventd.h"
 #include "util.h"
 #include "watchdogd.h"
+#include "vendor_init.h"
 
 using namespace std::string_literals;
 
@@ -207,6 +208,9 @@ static void restart_processes()
 }
 
 void handle_control_message(const std::string& msg, const std::string& name) {
+    if (!vendor_handle_control_message(msg, name))
+        return;
+
     Service* svc = ServiceManager::GetInstance().FindServiceByName(name);
     if (svc == nullptr) {
         LOG(ERROR) << "no such service '" << name << "'";
diff --git a/init/vendor_init.cpp b/init/vendor_init.cpp
index d3fd5ffe2..acc0e41fa 100644
--- a/init/vendor_init.cpp
+++ b/init/vendor_init.cpp
@@ -28,6 +28,7 @@ IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
 #include "vendor_init.h"
+#include <errno.h>
 
 /* init vendor override stubs */
 
@@ -35,3 +36,9 @@ __attribute__ ((weak))
 void vendor_load_properties()
 {
 }
+
+__attribute__ ((weak))
+int vendor_handle_control_message(const std::string& msg, const std::string& arg)
+{
+    return -ENOSYS;
+}
diff --git a/init/vendor_init.h b/init/vendor_init.h
index 9afb449be..d8ec073eb 100644
--- a/init/vendor_init.h
+++ b/init/vendor_init.h
@@ -27,7 +27,10 @@ OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
 IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+#include <string>
+
 #ifndef __INIT_VENDOR__H__
 #define __INIT_VENDOR__H__
 extern void vendor_load_properties(void);
+extern int vendor_handle_control_message(const std::string& msg, const std::string& arg);
 #endif /* __INIT_VENDOR__H__ */
-- 
2.17.0

